Use packaging.tags for doing compatible wheel tag calculation

chrahunt opened the pull request:
# Introduction

`packaging.tags` provides a simple [`sys_tags`](https://packaging.pypa.io/en/latest/tags/#packaging.tags.sys_tags) function for getting applicable tags for the running interpreter, but it does not allow customization of arguments. Since pip allows users to provide explicit interpreter (e.g. "pp"), platform (e.g. "linux_x86_64"), abi (e.g. "abi3"), and version (e.g. "35"), we have to use a different interface.

packaging.tags provides three functions for users that need to customize tag generation:

1. [`cpython_tags`](https://packaging.pypa.io/en/latest/tags/#packaging.tags.cpython_tags) - for tags applicable to CPython only
2. [`generic_tags`](https://packaging.pypa.io/en/latest/tags/#packaging.tags.generic_tags) - for tags applicable to any non-CPython Python implementation
3. [`compatible_tags`](https://packaging.pypa.io/en/latest/tags/#packaging.tags.compatible_tags) - for tags that are not specific to an interpreter
   implementation

`cpython_tags` and `generic_tags` are mutually exclusive, and return tags that are the highest priority. These tags are the most specific.

`compatible_tags` are always applicable, and a lower priority since they may just be compatible with e.g. the Python language version, but lack optimizations available in compiled code available for a specific interpreter/ABI/platform.

In this PR we remove most of our custom logic in `pip._internal.pep425tags`, instead relying on these three functions from `packaging.tags`.

# Approach

It's a given that we want to switch to `packaging.tags` for calculating compatible wheel tags, but while doing that we have a few other goals:

1. minimize bugs merged by making changes easy to review
2. if possible, compare the implementation in `packaging.tags` with what is currently in pip so we can understand possible issues that may be raised later

To satisfy 1, the PR is broken into small commits that should each be easy to recognize the literal changes of and confirm as equivalent or understand the scope of change. Each commit should pass lint checks and tests, which should reduce the number of trivial things that need to be worried about and ensure that all changes are self-contained.

To satisfy 2, we take an approach that transforms `pep425tags` into something that looks closer to `packaging.tags` before actually using the functions.

For each of the 3 custom tags functions, we make a copy of `get_supported`, then:

1. Refactor it locally, taking into account its new, more limited, responsibilities
2. Introduce the packaging.tags function for the case where there are no custom arguments provided
3. Customize arguments one-by-one and delegate to the packaging.tags function
4. When there is no pip-specific logic left, remove the intermediate function and use the packaging.tags function directly in `get_supported`

Following that, we remove unused utility methods, then in the end we're left with an implementation that relies solely on the tag generation in packaging.tags.

----

Fixes #6908.

pradyunsg commented:
You can probably do the following, to make the vendoring work properly.

```diff
diff --git a/tools/automation/vendoring/__init__.py b/tools/automation/vendoring/__init__.py
index 1c2ae3a6..a77e9e58 100644
--- a/tools/automation/vendoring/__init__.py
+++ b/tools/automation/vendoring/__init__.py
@@ -128,6 +128,7 @@ def vendor(ctx, vendor_dir):
     )
     remove_all(vendor_dir.glob('*.dist-info'))
     remove_all(vendor_dir.glob('*.egg-info'))
+    remove_all(vendor_dir.glob('*/py.typed'))
 
     # Cleanup setuptools unneeded parts
     (vendor_dir / 'easy_install.py').unlink()
```

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

pradyunsg commented on tools/automation/vendoring/__init__.py:
```diff
@@ -128,6 +128,7 @@ def vendor(ctx, vendor_dir):
     )
     remove_all(vendor_dir.glob('*.dist-info'))
     remove_all(vendor_dir.glob('*.egg-info'))
+    remove_all(vendor_dir.glob('*/py.typed'))
```
I'd prefer to keep this in a separate commit so that this isn't coupled with updating vendoring.

You probably thought of the same and might fix this when updating the vendored version -- I just wanted to flag that this is a thing we'd want to change before merging.

chrahunt commented:
Test failure would be fixed by #7395.

chrahunt commented on src/pip/_internal/pep425tags.py:
```diff
@@ -383,76 +173,46 @@ def get_supported(
-    if version is None:
-        version_info = get_impl_version_info()
-        versions = get_all_minor_versions_as_strings(version_info)
+    supported = []  # type: List[Union[Tag, Tuple[str, str, str]]]
```
Type can be simpler now that we're just using `packaging.tags`.

chrahunt commented:
Updated for latest commit, but had to update vendoring script locally to work around takluyver/flit#298.

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

chrahunt commented:
Rebased on master and updated to use the latest commit on master for `packaging`.

When `packaging` is released, I can start breaking pieces of this off into separate PRs.

chrahunt commented:
I broke several parts of this out into the following PRs:

* #7561 - Switch to packaging.tags.mac_version for macOS wheel tag backfilling
* #7562 - Use interpreter_name and interpreter_version from packaging.tags
* #7563 - Use packaging.tags.Tag to represent tag throughout the code
* #7566 - Extract platform list into separate function in pep425tags 

There will be a few trivial merge conflicts between them due to imports, but that shouldn't impact their reviewability.

Once those are done I will rebase this and it will be ready for review. At that point I think it would make sense to review the rest as one unit since the intermediate states don't really make sense on their own (though they should all pass lint checks and tests).

pradyunsg commented:
Thanks for breaking this up into smaller PRs! :)

xavfernandez commented:
Added to 20.0 milestone following https://github.com/pypa/pip/pull/7561#discussion_r363737980

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

pradyunsg commented on src/pip/_internal/pep425tags.py:
```diff
@@ -1,235 +1,37 @@
-    return sysconfig.get_config_var(var)
-
-
 def version_info_to_nodot(version_info):
```
follow up: I suggest moving the call to be changed -- changing `get_supported` to allow for version to be None (and use this value in that case).

pradyunsg commented:
Indeed. This was a decently large chunk of work which required a fair bit of figuring out the right way, and I know that you've spent time and effort to make the PRs easy to review!

Thanks @chrahunt for driving this and making it happen!
